woocommerce/google-listings-and-ads
# API PULL - Refactor Notification Jobs

### Changes proposed in this Pull Request:

Refactor Notifications Jobs and Tests for reducing duplications.

After emerging the work related to coupons notifications this PR does:

- Reactor all the generic methods shared between `SettingsNotificationsJob`, `ShippingNotificationsJob`, `CouponNotificationJob`,  `ProductNotificationJob`, into an abstract `AbstractNotificatioJob` class.
- Refactor `CouponNotificationJob` and `ProductNotificationJob` (Job using Item IDs)  into an abstract `AbstractNotificatioItemJob` 
- Refactor the tests (also using abstract tests)

⚠️  Some methods and classes (mostly regarding SyncerHooks, Helpers and MetaHandlers) can be refactored too. But are out scope of the Project. This might be done after adopting the new API PULL strategy and deprecating the current MC methods completely.

⚠️  IF your domain is not authorized for using the sandbox, this PR wont work as a successful notification it's needed to update the statuses. So one option is to hardcode the next line so the notification is always successful.

`NotificationsService.php`
```php
public function notify( string $topic, $item_id = null ): bool {
     return true;
}
```

### Detailed test instructions:


1. Create a product -- See `gla/jobs/notifications/products/process_item with the product and the `product.create `topic. Run it
2. Update the product -- See `gla/jobs/notifications/products/process_item with the product and the `product.update `topic. Run it
3. Delete the product -- See `gla/jobs/notifications/products/process_item with the product and the `product.delete `topic. Run it
4. Restore the product. See `gla/jobs/notifications/products/process_item with the product and the `product.create `topic
5. Create a coupon -- See `gla/jobs/notifications/coupons/process_item with the coupon and the `coupon.create `topic. Run it
6. Update the coupon -- See `gla/jobs/notifications/coupons/process_item with the coupon and the `coupon.update `topic. Run it
7. Delete the coupon -- See `gla/jobs/notifications/coupons/process_item with the coupon and the `coupon.delete `topic. Run it
8. Update the WooCommerce general Settings -- See `gla/jobs/notifications/settings/process_item without ID and the `settings.update` topic.
9.  Update/Delete/Create any WooCommerce Shipping -- See `gla/jobs/notifications/shipping/process_item without ID and the `settings.update` topic.
